### PR TITLE
CNV Live migration of VMIs

### PR DESCRIPTION
**Pull request by aburdenthehand**
This PR covers the following stories:
CNV-757 (Basic live migration)
CNV-759 (Progress/monitoring)
CNV-761(Cancel migration)
CNV-1495 (UI for above) 

No changes to the topic map as yet as migrating 1.4 content and assembling the CNV User Guide will begin next week. 

Preview content hosted here:
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-live-migration.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-live-migration-limits.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-migrate-vmi.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-monitor-vmi-migration.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-cancel-vmi-migration.html

### CODE REVIEW

File: modules/cnv-cancelling-vm-migration-web.adoc [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Module included in the following assemblies:
+//
+// * cnv_users_guide/cnv-cancel-vmi-migration.adoc
+
+[id="cnv-cancelling-vm-migration-web_{context}"]
+= Cancelling live migration of a virtual machine instance in the web console
+
+A live migration of the virtual machine instance can be cancelled using the 
+&#8942; button found on each virtual machine in the *Workloads* -> *Virtual Machines*
```
Comment (suomiy) [inline modules/cnv-cancelling-vm-migration-web.adoc]: kebap button (&#8942;) 

might be more descriptive

File: modules/cnv-configuring-vmi-live-migration-cli.adoc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+eviction strategy 
+
+The `LiveMigrate` eviction strategy ensures that a virtual machine instance is 
+live migrated to another node in the case of a node being placed in 
+maintenance mode, or if the `node drain` command is performed on the node. 
+
+Only custom virtual machines need to have this feature manually configured. This
+ eviction strategy is configured by default in virtual machine created from 
+common templates, or if a guest operating system is selected in either the 
+*Create Virtual Machine* or *Create Virtual Machine Template* wizards in the
```
Comment (suomiy) [inline modules/cnv-configuring-vmi-live-migration-cli.adoc]: This is thanks to the templates - we do not configure eviction strategy explicitly

**Review by mareklibra — COMMENTED** [review]
lgtm from the UI perspective considering notes above

File: modules/cnv-configuring-vmi-live-migration-cli.adoc (hunk 2 above)
Comment (aburdenthehand) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: Sorry @suomiy, I'm not sure I understand. Can you please elaborate?

Comment (suomiy) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: The VM is always created from a template - so the eviction strategy is not because of the guest operating system.

Comment (aburdenthehand) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: @suomiy I removed the part about guest operating systems. Does that make more sense for you?

File: modules/cnv-cancelling-vm-migration-web.adoc (hunk 1 above)
Comment (aburdenthehand) [reply modules/cnv-cancelling-vm-migration-web.adoc]: The 'kebab/kebap' menu conversation came up in OCP docs and I'm waiting to hear what the decision was. 
I'd be more inclined to call it an 'Actions' menu, since it mostly mirrors the options available for that button on the <Object> Description page

Comment (suomiy) [reply modules/cnv-cancelling-vm-migration-web.adoc]: +1 Action menu sounds good

File: modules/cnv-configuring-vmi-live-migration-cli.adoc (hunk 2 above)
Comment (suomiy) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: yes, looks ok

**Review by atiratree — APPROVED** [review]

File: modules/cnv-initiating-vm-migration-cli.adoc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+----
+
+. Create the object in the cluster:
++
+----
+oc create -f vmi-migrate.yaml
+----
+
+The *VirtualMachineInstanceMigration* object triggers a live migration of the 
+virtual machine instance and exists until the migration is complete.
```
Comment (dshchedr) [inline modules/cnv-initiating-vm-migration-cli.adoc]: As I know, *VirtualMachineInstanceMigration* object will exist all time while VMI exists (it will be removed only after stopping/deleting VM)

**Review by dshchedr — COMMENTED** [review]
Lgtm! Just one comment:

File: modules/cnv-configuring-live-migration-limits.adoc [hunk 4]
```diff
@@ -0,0 +1,40 @@
+
+// Module included in the following assemblies:
+//
+// cnv_users_guide/cnv-live-migration-limits.adoc
+
+[id="cnv-configuring-live-migration-limits_{context}"]
+= Configuring live migration limits
```
Comment (vladikr) [inline modules/cnv-configuring-live-migration-limits.adoc]: I'm not sure about the word `limits`. There can be different options, not just limiting factors. That's minor though...

File: modules/cnv-initiating-vm-migration-cli.adoc (hunk 3 above)
Comment (vladikr) [reply modules/cnv-initiating-vm-migration-cli.adoc]: Yes. Live migration object will remain as long as the VMI object exist or if manually deleted. In fact, there can be multiple migration object for each VMI.

File: modules/cnv-live-migration-limits-reftable.adoc [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+|2
+
+|`bandwidthPerMigration`
+|Bandwidth limit of each migration, in MiB/s.
+|64Mi
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
+per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has 
+not completed migration in 4800 seconds.
```
Comment (vladikr) [inline modules/cnv-live-migration-limits-reftable.adoc]: In case of a block migration, the size of the disks that are being migrated will be included in the calculation.

**Review by vladikr — COMMENTED** [review]
Looks good, thanks. Just a few minor comments.

File: modules/cnv-initiating-vm-migration-cli.adoc (hunk 3 above)
Comment (aburdenthehand) [reply modules/cnv-initiating-vm-migration-cli.adoc]: Updated to read: "The *VirtualMachineInstanceMigration* object triggers a live migration of the
virtual machine instance. This object exists in the cluster for as long as the
virtual machine instance is running, unless manually deleted."

File: modules/cnv-configuring-live-migration-limits.adoc (hunk 4 above)
Comment (aburdenthehand) [reply modules/cnv-configuring-live-migration-limits.adoc]: I've added "and timeouts" where appropriate.

**Review by vladikr — COMMENTED** [review]
:+1: Looks good to me after the changes. Thank you!

**Comment by dshchedr** [comment]
lgfm, thanks

File: modules/cnv-live-migration-limits-reftable.adoc (hunk 5 above)
Comment (aburdenthehand) [reply modules/cnv-live-migration-limits-reftable.adoc]: Now reads:
"The migration will be aborted if it has not completed in this time, in seconds
per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has
not completed migration in 4800 seconds. If the `Migration Method` is
`BlockMigration`, the size of the migrating disks is included in the calculation."

**Comment by vikram-redhat** [comment]
@openshift/team-documentation PTAL.

File: modules/cnv-initiating-vm-migration-web.adoc [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Module included in the following assemblies:
+//
+// cnv_users_guide/cnv-migrate-vmi.adoc
+
+[id="cnv-initiating-vm-migration-web_{context}"]
+= Initiating live migration of a virtual machine instance in the web console
+
+Migrate a running virtual machine instance to a different node in the cluster. 
+
+Migrate a virtual machine using the &#8942; button of the template in the
```
Comment (kalexand-rh) [inline modules/cnv-initiating-vm-migration-web.adoc]: I don't think you need this paragraph.

File: modules/cnv-initiating-vm-migration-web.adoc [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+Migrate a running virtual machine instance to a different node in the cluster. 
+
+Migrate a virtual machine using the &#8942; button of the template in the 
+*Workloads* -> *Virtual Machines* list, or using the *Actions* control of the 
+*Virtual Machine Details* screen.
+
+[NOTE]
+====
+Only admin users can initiate a virtual machine migration, although the
```
Comment (kalexand-rh) [inline modules/cnv-initiating-vm-migration-web.adoc]: I'd swap the clause order and consider making this a prereq instead of a note.

File: modules/cnv-initiating-vm-migration-web.adoc [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+[NOTE]
+====
+Only admin users can initiate a virtual machine migration, although the 
+*Migrate Virtual Machine* action is visible to all users.
+====
+
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
+. Click the &#8942; button of the virtual machine to migrate and select
```
Comment (kalexand-rh) [inline modules/cnv-initiating-vm-migration-web.adoc]: My suggestions for these alternate paths are the same as in the other module: I'd pull one, but if you keep them both, please restructure the step.

File: modules/cnv-live-migration-limits-reftable.adoc [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+// Module included in the following assemblies:
+//
+// cnv_users_guide/cnv-live-migration-limits.adoc
+
+[id="cnv-live-migration-limits-ref_{context}"]
+= Cluster-wide live migration limits and timeouts
+
+.Migration parameters
+|===
+|Parameter |Description |Default
```
Comment (kalexand-rh) [inline modules/cnv-live-migration-limits-reftable.adoc]: Default value

If you have information about acceptable ranges, consider adding another column to describe the values.

File: modules/cnv-live-migration-limits-reftable.adoc [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+|`parallelOutboundMigrationsPerNode`
+|Maximum number of outbound migrations per node.
+|2
+
+|`bandwidthPerMigration`
+|Bandwidth limit of each migration, in MiB/s.
+|64Mi
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds
```
Comment (kalexand-rh) [inline modules/cnv-live-migration-limits-reftable.adoc]: not "abort"

> :no_entry_sign: abort v
Do not use. Use “cancel,” “stop,” or “end."

File: modules/cnv-live-migration-limits-reftable.adoc [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+|Maximum number of outbound migrations per node.
+|2
+
+|`bandwidthPerMigration`
+|Bandwidth limit of each migration, in MiB/s.
+|64Mi
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
+per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has
```
Comment (kalexand-rh) [inline modules/cnv-live-migration-limits-reftable.adoc]: I'd mention the value of completionTimeoutPerGiB here. It's implied that it's the default value, but I've seen people file bugs about similar issues.

I'd also consider removing the future tense. I think it's clear enough in present tense.

File: modules/cnv-live-migration-limits-reftable.adoc [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
+per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has 
+not completed migration in 4800 seconds. If the `Migration Method` is 
+`BlockMigration`, the size of the migrating disks is included in the calculation. 
+|800
+
+|`progressTimeout`
+|The migration will be aborted if memory copy fails to make progress in this
```
Comment (kalexand-rh) [inline modules/cnv-live-migration-limits-reftable.adoc]: not "abort"

File: modules/cnv-monitoring-vm-migration-cli.adoc [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+The status of the virtual machine migration is stored in the *Status* component 
+of the *VirtualMachineInstance* configuration. 
+
+.Procedure
+
+* Use the `oc describe` command on the migrating virtual machine instance:
+
+----
+oc describe vmi vmi-fedora
```
Comment (kalexand-rh) [inline modules/cnv-monitoring-vm-migration-cli.adoc]: $

File: modules/cnv-monitoring-vm-migration-web.adoc [hunk 14]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// cnv_users_guide/cnv-monitor-vmi-migration.adoc
+
+[id="cnv-monitoring-vm-migration-web_{context}"]
+= Monitoring live migration of a virtual machine instance in the web console
+
+For the duration of the migration, the virtual machine will have a status of
```
Comment (kalexand-rh) [inline modules/cnv-monitoring-vm-migration-web.adoc]: s/will have/has

File: modules/cnv-monitoring-vm-migration-web.adoc [hunk 15]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// cnv_users_guide/cnv-monitor-vmi-migration.adoc
+
+[id="cnv-monitoring-vm-migration-web_{context}"]
+= Monitoring live migration of a virtual machine instance in the web console
+
+For the duration of the migration, the virtual machine will have a status of 
+*Migrating*. This is displayed in the *Virtual Machines* list, or in the
```
Comment (kalexand-rh) [inline modules/cnv-monitoring-vm-migration-web.adoc]: s/This/This status

no comma

File: modules/cnv-monitoring-vm-migration-web.adoc [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+[id="cnv-monitoring-vm-migration-web_{context}"]
+= Monitoring live migration of a virtual machine instance in the web console
+
+For the duration of the migration, the virtual machine will have a status of 
+*Migrating*. This is displayed in the *Virtual Machines* list, or in the 
+*Virtual Machine Details* screen for the migrating virtual machine. 
+
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
```
Comment (kalexand-rh) [inline modules/cnv-monitoring-vm-migration-web.adoc]: Same as before, I'd pick one, but if you keep them both, please reformat the steps.

Since there's one step, lead with * instead of .

**Review by kalexand-rh — COMMENTED** [review]
I see a couple of style changes and have some suggestions. Overall, it's looking good!

File: modules/cnv-cancelling-vm-migration-web.adoc [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+= Cancelling live migration of a virtual machine instance in the web console
+
+A live migration of the virtual machine instance can be cancelled using the _Actions_ 
+&#8942; button found on each virtual machine in the *Workloads* -> *Virtual Machines* 
+list, or from the *Actions* control button on the *Virtual Machine Details* screen.
+
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
+. Click the Actions &#8942; button of the virtual machine and select
```
[the remark this answers is not in the document]
Comment (aburdenthehand) [reply modules/cnv-cancelling-vm-migration-web.adoc]: I think there's benefit to having both methods. How about:
. You can cancel the migration from this screen, which makes it easier to perform actions on multiple virtual machines in the one screen, or from the *Virtual Machine Details* screen:
** Click :kebab: at the end of  virtual machine and select           
*Cancel Virtual Machine Migration*.
** Click the virtual machine name to open the *Virtual Machine Details*               
screen and click *Actions* -> *Cancel Virtual Machine Migration*.
?
It's a bit wordy, but it offers a 'why?' for the first method - the second method should be self-explanatory.

File: modules/cnv-configuring-live-migration-limits.adoc [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+.Procedure
+
+* Edit the `kubevirt-config` configuration file and add the necessary 
+live migration parameters. The following example shows the defaults:
++
+----
+oc edit configmap kubevirt-config -n kubevirt
+----
++
+[source,yaml]
```
[the remark this answers is not in the document]
Comment (aburdenthehand) [reply modules/cnv-configuring-live-migration-limits.adoc]: The reference table is the next module in the assembly. I'd prefer a link as well but ...

File: modules/cnv-cancelling-vm-migration-web.adoc (hunk 17 above)
[the remark this answers is not in the document]
Comment (kalexand-rh) [reply modules/cnv-cancelling-vm-migration-web.adoc]: That works! It's good to have the reason why you'd pick the other one. I'd apply the same style to the other similar instances.

File: modules/cnv-configuring-vmi-live-migration-cli.adoc [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+live migrated to another node in the case of a node being placed in 
+maintenance mode, or if the `node drain` command is performed on the node. 
+
+Only custom virtual machines need to have this feature manually configured. This
+ eviction strategy is configured by default in virtual machines created from 
+common templates. 
+
+.Prerequisites
+
+* The live migration feature must be enabled in the cluster.
```
[the remark this answers is not in the document]
Comment (aburdenthehand) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: Right you are. This is for #14923 (node maintenance). I mistakenly added this instead of the live migration concept that is shared across both that PR and this one. I've fixed that now, and will add the feedback you've noted for this topic in the 14923.

File: modules/cnv-configuring-vmi-live-migration-cli.adoc [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+.Procedure
+
+* Add the `evictionStrategy: LiveMigrate` to the `spec` in the virtual machine instance 
+configuration file. The following example snippet shows the eviction strategy 
+configured in the *_custom-vmi_* configuration file:
```
[the remark this answers is not in the document]
Comment (aburdenthehand) [reply modules/cnv-configuring-vmi-live-migration-cli.adoc]: It's just an example VMI config so I've changed the procedure to use `$ oc edit vmi ...` for the example

**Comment by aburdenthehand** [comment]
@kalexand-rh - Thanks for the review Kathryn. I have updated the PR as necessary.
I added a new file cnv-understanding-live-migration.adoc that is shared with another PR (as discussed). If it passes muster then I believe this is ready to merge.

**Comment by kalexand-rh** [comment]
I scanned the final PR, and it looks fine. Merging.
